### PR TITLE
Implement separate types for the CUDA and HIP/ROCm backends

### PR DESCRIPTION
**Pull request by fwyzard**
The `UniformCudaHip` types are now templated on a new traits-like `struct` that encapsulates the CUDA or HIP API, depending on the backend being used.

Users should use the explicit backend types like `DevHipRt`, `AccGpuHipRt`, ... and `DevCudaRt`, `AccGpuCudaRt`, ...

### CODE REVIEW

**Comment by fwyzard** [comment]
The header files and namespaces could probably be arranged better.
For example:
  - move all `*UniformCudaHipRt*.hpp` files to a `.../cuda_hip/` subdirectory
  - move all `*UniformCudaHipRt*` types to the `uniform_cuda_hip` or `detail` namespace

**Comment by fwyzard** [comment]
The random number generators are not yet ported to the new interface -- hence many of the failures.

File: include/alpaka/acc/AccGpuCudaRt.hpp
Comment (psychocoderHPC) [inline include/alpaka/acc/AccGpuCudaRt.hpp:20]: We started to discuss this line [here](https://github.com/fwyzard/alpaka/commit/34f251aec7be8565a081aecc872467d92e0182af#r69328577) already.

summery: removing the struct would give the wrong naming via `GetAccName` and type via `AccType`

Comment (psychocoderHPC) [reply include/alpaka/acc/AccGpuCudaRt.hpp:20]: @fwyzard I suggest to keep the implementation with the struct instead of the alias to keep the separation of the HIP and CUDA accelerator.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: include/alpaka/acc/AccGpuCudaRt.hpp
Comment (fwyzard) [reply include/alpaka/acc/AccGpuCudaRt.hpp:20]: > removing the struct would give the wrong naming via `GetAccName` and type via `AccType`

Actually it doesn't, because `GetAccName` returns a name based on `API::name`:
```c++
        template<typename API, typename TDim, typename TIdx>
        struct GetAccName<AccGpuUniformCudaHipRt<API, TDim, TIdx>>
        {
            ALPAKA_FN_HOST static auto getAccName() -> std::string
            {
                return "AccGpu" + API::name + "Rt<" + std::to_string(TDim::value) + "," + typeid(TIdx).name() + ">";
            }
        };
 ```

And `AccType` should be `AccGpuUniformCudaHipRt<API, ...>` which is the same type as either `AccGpuCudaRt<...>` or `AccGpuHipRt<...>`.

**Comment by fwyzard** [comment]
Rebased, and fixed all tests failing locally with gcc and cuda.

**Comment by fwyzard** [comment]
@psychocoderHPC @bernhardmgruber I'm not sure how to handle this warning treated as error:
```
In file included from /home/runner/work/alpaka/alpaka/build/test/analysis/headerCheck/src/acc/AccGpuCudaRt.hpp.cpp:1:
In file included from /home/runner/work/alpaka/alpaka/include/alpaka/acc/AccGpuCudaRt.hpp:15:
/home/runner/work/alpaka/alpaka/include/alpaka/core/ApiCudaRt.hpp:16:41: error: declaration requires an exit-time destructor [-Werror,-Wexit-time-destructors]
        static inline const std::string name = "Cuda";
                                        ^
```

GCC does not complain abut it... if it is an issue I can certainly work around it, but I'd like to hear your opinion about it.

**Comment by fwyzard** [comment]
OK, I'll just change the `std::string` to a `const char[]` buffer.

**Comment by bernhardmgruber** [comment]
> OK, I'll just change the `std::string` to a `const char[]` buffer.

`std::string_view` should also work.

**Comment by fwyzard** [comment]
As discussed this morning, a test for this functionality is available at https://github.com/fwyzard/test-rocm-cuda-alpaka .

With the current `develop` branch (actually with #1678 to make the explicit types available), the test fails with
```
ROCm runtime initialised
CUDA runtime initialised
CPU-only serial runtime initialised

device: Radeon Pro WX 9100
accelerator: AccGpuHipRt<1,j>
result: 10
ROCm kernel ran successfully

device: NVIDIA GeForce GTX 1080 Ti
accelerator: AccGpuCudaRt<1,j>
result: 10
terminate called after throwing an instance of 'std::runtime_error'
  what():  alpaka/include/alpaka/mem/buf/BufUniformCudaHipRt.hpp(249) 'hipFree(ptr)' returned error  : 'hipErrorInvalidValue': 'hipErrorInvalidValue'!
Aborted (core dumped)
```

With this #1665 (this PR), the test passes with
```
ROCm runtime initialised
CUDA runtime initialised
CPU-only serial runtime initialised

device: Radeon Pro WX 9100
accelerator: AccGpuHipRt<1,j>
result: 10
ROCm kernel ran successfully

device: NVIDIA GeForce GTX 1080 Ti
accelerator: AccGpuCudaRt<1,j>
result: 10
CUDA kernel ran successfully

device: AMD Ryzen 9 5900X 12-Core Processor            
accelerator: AccCpuSerial<1,j>
result: 10
CPU-only serial kernel ran successfully
```

File: include/alpaka/acc/AccGpuUniformCudaHipRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/acc/AccGpuUniformCudaHipRt.hpp]: Should we keep the name `API` or name it `Api` or `TApi`?

File: include/alpaka/core/ApiCudaRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/core/ApiCudaRt.hpp]:
```suggestion
        inline static constexpr auto success = ::cudaSuccess;
```
This only makes a difference if the constant is ODR-used anywhere (e.g. its address is taken). I would add it though in case someone would like to do that. Applies to all constants in this struct and in `ApiHipRt`.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/core/ApiCudaRt.hpp]:
```suggestion
        static inline cudaError_t deviceReset()
```
I personally dislike this relic from C. But this is not blocking this PR.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/core/ApiCudaRt.hpp]:
```suggestion
        static inline cudaError_t deviceSetLimit(cudaLimit limit, size_t value)
```

File: include/alpaka/core/ApiCudaRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/core/ApiCudaRt.hpp]:
```suggestion
        static inline cudaError_t funcGetAttributes(cudaFuncAttributes* attr, const void* func)
```
We don't need the elaborated type specifier here. Applies to other places as well.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/core/ApiCudaRt.hpp]: If this is a device function, don't we need `__device__`?

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Do we even need the #else branch anymore then?

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Can we remove the preprocessor directives completely?

File: include/alpaka/mem/buf/BufUniformCudaHipRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/mem/buf/BufUniformCudaHipRt.hpp]: But if both backends are enabled, you would need both includes?
```suggestion
#        include <alpaka/core/Cuda.hpp>
#        include <alpaka/core/Hip.hpp>
struct ApiCudaRt;
struct ApiHipRt;
```

File: include/alpaka/mem/buf/uniformCudaHip/Set.hpp
Comment (bernhardmgruber) [inline include/alpaka/mem/buf/uniformCudaHip/Set.hpp]: Why did you readd this check? The `view` is zero dimensional, it must have a width of 1.

File: include/alpaka/pltf/PltfUniformCudaHipRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/pltf/PltfUniformCudaHipRt.hpp]: Just include both headers. No need for the `#if`.

File: include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp
Comment (bernhardmgruber) [inline include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp]:
```suggestion
```

File: include/alpaka/acc/AccGpuUniformCudaHipRt.hpp
Comment (fwyzard) [reply include/alpaka/acc/AccGpuUniformCudaHipRt.hpp]: Anything works for me.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp]: Right... I copied the signatures from the CUDA header, but you are right they can be simplified for the C++-only case (like Alpaka).

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
Comment (fwyzard) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Yes, that's definitely a leftover from some test !

File: include/alpaka/core/ApiCudaRt.hpp
Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp]: According to [cppreference.com](https://en.cppreference.com/w/cpp/language/constexpr)
> A `constexpr` specifier used in a function or `static` data member (since C++17) declaration implies `inline`.

so the explicit `inline` should not be necessary.

I can still add it if you prefer, though.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp]: No, they are actually host functions - I don't remember where I copied the (wrong) comment from.

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
Comment (fwyzard) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp]: Nope.

File: include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp
Comment (fwyzard) [reply include/alpaka/kernel/TaskKernelGpuUniformCudaHipRt.hpp:374]: Yes.

File: include/alpaka/mem/buf/uniformCudaHip/Set.hpp
Comment (fwyzard) [reply include/alpaka/mem/buf/uniformCudaHip/Set.hpp]: I had to rebase my changes on top of the latest developments, it may have slipped in while resolving the conflicts.

File: include/alpaka/mem/buf/BufUniformCudaHipRt.hpp
Comment (fwyzard) [reply include/alpaka/mem/buf/BufUniformCudaHipRt.hpp]: Yes.
Also, the forward declarations should be in the `alpaka` namespace - I'll fix them.

Comment (fwyzard) [reply include/alpaka/mem/buf/BufUniformCudaHipRt.hpp]: Well, it's actually not possible to enable both backends at the same time, because of the conflicts between `cuda_runtime.h` and `hip/hip_runtime.h` :-/

File: include/alpaka/core/ApiCudaRt.hpp
Comment (bernhardmgruber) [reply include/alpaka/core/ApiCudaRt.hpp]: Cool! I learned sth! You can omit the `inline` then.

**Comment by fwyzard** [comment]
@bernhardmgruber I've rebased the branch and addressed all simple comments.

The only one left is whether to use `API` or `Api` or `TApi` as the template parameter name (I'm fine with any option, just let me know what you all prefer).

**Comment by bernhardmgruber** [comment]
> @bernhardmgruber I've rebased the branch and addressed all simple comments.
> 
> The only one left is whether to use `API` or `Api` or `TApi` as the template parameter name (I'm fine with any option, just let me know what you all prefer).

OK, let's go with `TApi` then.
The rest of the PR LGTM.

**Comment by fwyzard** [comment]
Latest set of changes:
  - renamed `API` to `TApi` everywhere
  - changed the two "policy" structs to use the generic types (_e.g._ `Stream_t`) instead of the CUDA/HIP ones (_e.g._ `cudaStream_t`/`hipStream_t`) or just `auto`
  - rebased on top of daa516b5fcfd110bf1ba97a4bc3a4d2b9af1ce3b

**Comment by fwyzard** [comment]
8d0d179a6af8d258eb31d075965d6879a13a7869 is tricker to reimplement... ideally I would like to wrap the difference between `cudaLaunchHostFunc` and `cudaStreamAddCallback` directly in `ApiCudaRt.hpp` -- but I'm not yet sure how to do it.

**Comment by fwyzard** [comment]
The last commit ec07d47d1512ea780042893167adbaaf7fdcb373 reimplements the changes from 8d0d179a6af8d258eb31d075965d6879a13a7869 with a somewhat different approach.

All differences between CUDA versions and between CUDA and HIP are encapsulated in `ApiCudaRt.hpp` and `ApiHipRt.hpp`:
  - CUDA versions < 10.0 use `cudaStreamAddCallback`
  - CUDA versions >= 10.0 use `cudaLaunchHostFunc`
  - all HIP versions so far use `hipStreamAddCallback`
  - CUDA uses the `__stdcall` calling convention on Windows (CUDART_CB)
  - HIP uses the standard calling convention on all platforms

The interface exposed by both APIs is similar to that of `cudaLaunchHostFunc`, without the use of a special calling convention.

**Comment by fwyzard** [comment]
@j-stephan during the VC yesterday we agreed that this PR should target 0.10.0/1.0.0, as it is too invasive for 0.9.0 .

**Review by bernhardmgruber — DISMISSED** [review]

**Review by j-stephan — DISMISSED** [review]
Thanks for the PR. I want to push out the RC for 0.9.0 this afternoon so I'm locking the merge button until then.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (j-stephan) [inline include/alpaka/core/ApiCudaRt.hpp]:
```suggestion
        static inline Error_t freeAsync([[maybe_unused]] void* devPtr, [[maybe_unused]] Stream_t stream)
        {
#    if BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 2, 0)
            return ::cudaFreeAsync(devPtr, stream);
#    else
            // Not implemented.
            return errorUnknown;
#    endif
```

File: include/alpaka/core/ApiCudaRt.hpp
Comment (j-stephan) [inline include/alpaka/core/ApiCudaRt.hpp]:
```suggestion
        static inline Error_t mallocAsync([[maybe_unused]] void** devPtr, [[maybe_unused]] size_t size, [[maybe_unused]] Stream_t stream)
        {
#    if BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 2, 0)
            return ::cudaMallocAsync(devPtr, size, stream);
#    else
            // Not implemented.
            return errorUnknown;
#    endif
        }
```

File: include/alpaka/core/ApiHipRt.hpp
Comment (j-stephan) [inline include/alpaka/core/ApiHipRt.hpp]:
```suggestion
        static inline Error_t deviceSetLimit(Limit_t, size_t)
        {
            // Not implemented.
            return errorUnknown;
        }
```

File: include/alpaka/core/ApiHipRt.hpp
Comment (j-stephan) [inline include/alpaka/core/ApiHipRt.hpp]:
```suggestion
        static inline Error_t freeAsync(void*, Stream_t)
        {
            // Not implemented.
            return errorUnknown;
        }
```

File: include/alpaka/core/ApiHipRt.hpp
Comment (j-stephan) [inline include/alpaka/core/ApiHipRt.hpp]:
```suggestion
        static inline Error_t mallocAsync(void**, size_t, Stream_t)
        {
            // Not implemented.
            return errorUnknown;
        }
```

**Review by j-stephan — CHANGES_REQUESTED** [review]
Thanks! A few minor suggestions:

File: include/alpaka/acc/AccGpuUniformCudaHipRt.hpp
Comment (j-stephan) [reply include/alpaka/acc/AccGpuUniformCudaHipRt.hpp]: `TApi` is the way.

File: include/alpaka/core/ApiCudaRt.hpp
Comment (fwyzard) [inline include/alpaka/core/ApiCudaRt.hpp]: I see... how would you see this approach instead ?
```suggestion
#    if BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 2, 0)
        static inline Error_t freeAsync(void* devPtr, Stream_t stream)
        {
            return ::cudaFreeAsync(devPtr, stream);
        }
#    else
        static inline Error_t freeAsync(void*, Stream_t)
        {
            // Not implemented.
            return errorUnknown;
        }
#    endif
```

Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp]: @j-stephan

Comment (j-stephan) [reply include/alpaka/core/ApiCudaRt.hpp]: Yes, that would work. Is there a reason you don't want to use `[[maybe_unused]]`?

Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp]: In general I'm fine with using `[[maybe_unused]]`, for example when the conditions when a parameter is used or not are not obvious.

In this specific case we know when the parameters are used or not, so I though this approach would make more sense.

But, all in all, I don't mind much, and I'm fine using whatever you prefer.

Comment (j-stephan) [reply include/alpaka/core/ApiCudaRt.hpp]: > In this specific case we know when the parameters are used or not, so I though this approach would make more sense.

Sure, but `[[maybe_unused]]` is a compiler hint ;-) We also know this when using `if constexpr`, for example. For that reason I prefer `[[maybe_unused]]`.

Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp]: OK.

**Comment by fwyzard** [comment]
Rebased on top of `develop`, and addressed the last round of comments.

File: include/alpaka/core/ApiHipRt.hpp
Comment (fwyzard) [reply include/alpaka/core/ApiHipRt.hpp]: I've left the parameter names in `/* comments */`, to keep a reminder of what they are supposed to be, when the functions will eventually be supported by HIP.

**Review by j-stephan — DISMISSED** [review]

**Review by bernhardmgruber — DISMISSED** [review]

File: include/alpaka/core/ApiCudaRt.hpp
Comment (psychocoderHPC) [inline include/alpaka/core/ApiCudaRt.hpp:10]: Please add the license header to this file:

```
/* Copyright 2022 Andrea Bocci
 *
 * This file is part of alpaka.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
```

We add the license to each header, in cases someone is coping one file out of alpaka the license is still clear.

File: include/alpaka/core/ApiHipRt.hpp
Comment (psychocoderHPC) [inline include/alpaka/core/ApiHipRt.hpp:10]: Please add the license header to this file:

```
/* Copyright 2022 Andrea Bocci
 *
 * This file is part of alpaka.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
```

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]
I found only two files without a license. All other code looks fine.

Thanks for this nice PR!

File: include/alpaka/core/ApiHipRt.hpp
Comment (fwyzard) [reply include/alpaka/core/ApiHipRt.hpp:10]: :+1:

File: include/alpaka/core/ApiCudaRt.hpp
Comment (fwyzard) [reply include/alpaka/core/ApiCudaRt.hpp:10]: :+1:

**Review by bernhardmgruber — APPROVED** [review]

**Review by psychocoderHPC — APPROVED** [review]

**Comment by fwyzard** [comment]
:+1:

**Pull request merged**
